lib/gems/pending/util/miq_ftp_lib.rb

```diff
@@ -0,0 +1,70 @@
+require 'net/ftp'
+require 'logger'
+
+# Helper methods for net/ftp based classes and files.
+#
+# Will setup a `@ftp` attr_accessor to be used as the return value for
+# `.connect`, the main method being provided in this class.  Will also setup
+# logging if not already done for the particular class (that follows the
+# VmdbLogger conventions, and setup a `uri` attr_accessor if that doesn't
+# already exist.
+module MiqFtpLib
+  def self.included(klass)
+    klass.send(:attr_accessor, :ftp)
+
+    klass.send(:attr_accessor, :uri) unless klass.instance_methods.include?(:uri=)
+
+    unless klass.instance_methods.include?(:_log)
+      klass.send(:define_method, :_log) do
+        self.class.instance_logger
```

Comment:
This makes me feel like this would be better as a class so you could assign a logger rather than a module where we need to assume some methods exist.

I'm not sure of the full direction for this so maybe it makes more sense with the full picture, but I still think it's worth considering.

Reply:
> I'm not sure of the full direction for this so maybe it makes more sense with the full picture, but I still think it's worth considering.

Okay, I have been distracted between a few things, and this response has been half completed on my computer for about 3 hours, I will give a "brief" summary of what I am planning:

* I am in the process of creating a `MiqFileStorage` and `MiqFileStorage::Interface` which is meant to be the base class for `MiqGenericMountSession` classes, and what I am currently going to call "`MiqUploadTargets`" (encompassing `FTP`, `s3`, `Swift`, etc.).
* `MiqFileStorage::Interface` will be in charge of defining top level methods and underlying implementation methods can be overwritten by subclasses per implementation (where the input from a IO stream goes when writing to a split file)
* `MiqFileStorage` will be roughly what `with_mount_session` is at the moment in `EvmDatabaseOps` in `ManageIQ/manageiq`.
* I plan on putting `FTP` and `s3` integrations under a `UploadTarget` (working title), which will be managed by `MiqFileStorage`, but conform to the `MiqFileStorage::Interface`
* the `FtpUploadTarget` will want to probably share the connection logic of `FileDepotFtp`, and I see this as the easiest way to do it.

As far as the specifics in your question:

> This makes me feel like this would be better as a class so you could assign a logger rather than a module where we need to assume some methods exist.

I am basically trying to keep `FileDepotFtp` in as much of it's current state without changing it, while being able to share it's logic without calling out to it directly.  The `_log` statements are part of that issue, and this is the best way I could think of solving it.  The `instance_logger` is the exact same interface that is used under the covers for `VmdbLogger` when defining `_log`, so this is basically me ripping that implementation off.

This most of the `if` logic here is to account for classes that are not `FileDepotFtp`, and not the other way around.

Reply:
I'm going to move the general design discussion to a regular (non-code specific) comment